### DISCUSSION TITLE
libc used in gramine

### DISCUSSION THREAD

**Discussion by blossomin** [gramineproject/gramine#1621] [General]
I found that Gramine uses a patched version libc, glibc or musl. In the building from the source tutorial, we need to install SGX SDK.  As far as I know, there is one tuned libc in the SDK. This specific libc doesn't support some functions, like printf, fopen. Then I'm wondering: if we don't consider the compatibility issue of this tuned libc, can we link our application directly against the SGX SDK libc rather than using the patched version? Of course, we may also need to patch the GRAMINE_SYSCALL for SDK libc. Just some random thought, not sure whether this works.

Another question about the subprojects used in Gramine libos: cJSON, curl, GCC, glibc, musl, mbedtls, tomlc, uthash.
I know musl and glibc are the c library used by the application linking.  Can you provide some ideas about where the other projects are used?

Best,
Blossomin

**Comment by dimakuv** [comment]
> As far as I know, there is one tuned libc in the SDK. This specific libc doesn't support some functions, like printf, fopen. Then I'm wondering: if we don't consider the compatibility issue of this tuned libc, can we link our application directly against the SGX SDK libc rather than using the patched version? Of course, we may also need to patch the GRAMINE_SYSCALL for SDK libc. Just some random thought, not sure whether this works.

We never tried it, but you are right. Theoretically, yes, you can link this SGX SDK libc into your application, and hopefully with some minimal tweaks and patches you'll be able to make it work.

Though I'm not sure -- what's the benefit of this? If you want something as small as possible, then build your app against Musl. This gives you small and robust libc implementation.

But in general -- yes, if you want to go to the extreme, feel free to link against SGX SDK libc (or maybe some other super-small libc like `newlib`). We'll be curious to know the results of your experiment.

> Another question about the subprojects used in Gramine libos: cJSON, curl, GCC, glibc, musl, mbedtls, tomlc, uthash.
I know musl and glibc are the c library used by the application linking. Can you provide some ideas about where the other projects are used?

- cJSON: used in RA-TLS/SecretProv libraries to parse JSON objects embedded in HTTPS requests (e.g. when communicating with Intel Attestation Service).
- curl: used in RA-TLS/SecretProv libraries to send/receive HTTPS requests/responses (e.g. when communicating with Intel Attestation Service).
- GCC: **not used**, where did you find this subproject? It's only used for building Gramine itself. **UPDATE 26. Oct 2023**: Sorry, indeed we have GCC as a subproject -- we don't build GCC from there but instead only the **libgomp** (OpenMP) library, which is patched to use Glibc wrapper `futex()` instead of the raw syscall.
- Glibc/Musl: you described already.
- mbedTLS: used for all crypto- and TLS-operations in Gramine and its tools:
   - crypto (`libmbedcrypto`) is used in Gramine core, for e.g. AES-GCM on Encrypted Files and SHA256 on Trusted Files.
   - TLS (`libmbedtls` and `libmbedx509`) is used in RA-TLS/SecretProv libraries, to establish TLS connections and to create/verify X.509 certificates.
- Tomlc99: used in Gramine core to parse the manifest file (which is written in the TOML syntax).
- UTHash: used in Gramine core, in particular in Encrypted Files, to create a Least Recently Used (LRU) cache of blocks of encrypted files, as a performance optimization.

**Reply by blossomin** [reply]
Thanks for your reply.

One further question about the libc:

do the gramine libos or pal use the SDK libc? In fact, I'm confused about the SGX SDK use in Gramine. In the native sgx develop workflow, developers need to specify enclave interfaces in EDL file. In gramine libos context, the "sgx application" should the pal loader, and the interfaces are fixed. And gramine has its own host_ECall.S and enclave_ECall.S file to handle Ecalls and Ocalls.  Then where is the SDK used in gramine libos and pal? And do the libos and pal use libc in the SDK? Maybe this question is too detailed, any idea is appreciated.

**Reply by dimakuv** [reply]
> do the gramine libos or pal use the SDK libc? In fact, I'm confused about the SGX SDK use in Gramine.

No, Gramine does **not** use Intel SGX SDK at all. Therefore, SDK libc is also **not used** in Gramine.

The fact that we ask to install Intel SGX SDK/PSW in our documentation is simply because *it is easier to explain to users*. In fact, the only thing that we need is Intel SGX PSW, and we only use PSW services that are needed for SGX remote attestation (like generating the SGX quote and verifying the SGX quote). The Intel SGX SDK services/packages are **not used** by Gramine at all.

Unfortunately, SGX SDK and SGX PSW are kinda conflated together (they have a single GitHub repository, a single set of documentation), so Gramine doesn't make a clear distinction between SDK and PSW in the documentation. But for advanced users: SGX PSW is used (for SGX attestation), SGX SDK is not used.

**Reply by blossomin** [reply]
Very helpful, thanks😊
